Peer 190802

thereseanders opened the pull request.

mine-cetinkaya-rundel commented on vignettes/peer.Rmd:
```diff
@@ -0,0 +1,249 @@
+This vignette introduces the peer review functionality of `ghclass` and provides a step-by-step guide. For more information on the pedagogy of peer review in undergraduate courses, see the draft of the [accompanying paper](https://github.com/thereseanders/cprpaper/blob/master/cetinkayarundelanders_cpr.pdf) by the package authors.
+
+# Before peer review
+The peer review process is initiated after a regular assignment has been distributed to students. Below, we run through a quick example of how to set up a fictional assignment "hw2" for four students using the `create_repo()` function. For a more detailed description of how to set up student repositories in a GitHub Organization,  see the `ghclass` vignette.
```
Link to vignette, referring to it as the "Get started" vignette

thereseanders replied:
Done

mine-cetinkaya-rundel commented on vignettes/peer.Rmd:
```diff
@@ -0,0 +1,249 @@
+repo_add_user(glue::glue("{org}/{prefix}{user}"), user)
+
+# 4) Place homework files in repositories
+task <- "/Users/thereseanders/Documents/RStudio/introggplot2/hw2/hw2_task.Rmd"
```
Is this real code or "fake" code? If latter can you edit the paths so it reads someting like "/Users/thereseanders/Teaching/introggplot2", in case the Documents/RStudio bit is confusing

thereseanders replied:
Done

mine-cetinkaya-rundel commented on vignettes/peer.Rmd:
````diff
@@ -0,0 +1,249 @@
+```{r, eval = F}
+library(ghclass)
+devtools::load_all()
+# 1) Set up vector of student user names, HW identifier prefix, and organization name
````
For enumeration let's use `1.` instead of `1)` -- I think this shows up in a few places

thereseanders replied:
done

mine-cetinkaya-rundel commented on vignettes/peer.Rmd:
````diff
@@ -0,0 +1,249 @@
+### Collect scores given by reviewers
+`peer_score_review()` collects the scores given by reviewers to authors. We recommend keeping the default setting of `write_csv = TRUE` to save the output as a `.csv` file in the working directory.
+```{r, eval = F}
+peer_score_review(
````
I like how you've named the arguments in the vignette, so much more clear!

mine-cetinkaya-rundel commented on vignettes/peer.Rmd:
````diff
@@ -0,0 +1,249 @@
+```
+
+### Double blind vs. single blind review
+The `dblind` parameter is used to determine whether authors know who reviewed their assignments (`dblind = FALSE`), or whether the entire review process is anonymous (`dblind = TRUE`). The default is for authors to know who reviewed their assignments. Keeping reviewers non-anonymous is likely to discourage overly harsh or rude reviews. Authors will always remain anonymous to reviewers.
````
Would it take up too much space to name the argument `double_blind()`? If not, let's go with that so it's clearer.

thereseanders replied:
changed to `double_blind`

mine-cetinkaya-rundel commented on vignettes/peer.Rmd:
````diff
@@ -0,0 +1,249 @@
+
+This parameter should be set once for each assignment and then passed to each of the functions in the return review process. `dblind` determines the naming of folders on authors repositories. Changing from a setting of double blind to single blind review, or vice versa, in the middle of the return review process will prevent `peer_add_file_aut()`, `peer_return()` and `peer_score_rating()` to loca ...
+```{r, eval = F}
+dblind = TRUE
````
I wonder if this should be an option for the package that's stored globally instead of a variable... What do you think?

thereseanders replied:
We discussed that we want users to set this variable actively to encourage them to think carefully whether they want reviewers to be anonymous to authors or not.

mine-cetinkaya-rundel commented on R/peer.R:
```diff
@@ -0,0 +1,909 @@
+#' @param m Numeric. Number of reviews per user. Must be larger than zero and smaller than the number of users.
+#' @param seed Numeric. Random seed for assignment, defaults to `12345`.
+#' @param write_csv Logical. Whether the roster data frame should be saved to a `.csv` file in the current working directory, defaults to TRUE.
+#'
```
We should make use of the `@family` tag for all the `peer_*()` functions. You might take a look at the implementation of the `usethis::pr_*()` functions for documenting a set of related functions. Also see https://r-pkgs.org/man.html#roxygen-comments.

mine-cetinkaya-rundel replied:
Or `@seealso`...

thereseanders replied:
Done.

mine-cetinkaya-rundel commented on vignettes/peer.Rmd:
````diff
@@ -0,0 +1,249 @@
+The peer review process is initiated after a regular assignment has been distributed to students. Below, we run through a quick example of how to set up a fictional assignment "hw2" for four students using the `create_repo()` function. For a more detailed description of how to set up student repositories in a GitHub Organization,  see the `ghclass` vignette.
+```{r, eval = F}
+library(ghclass)
+devtools::load_all()
````
For the vignette we shouldn't need `load_all()`

thereseanders commented on NAMESPACE:
```diff
@@ -1,5 +1,6 @@
 # Generated by roxygen2: do not edit by hand
 
+export("%>%")
```
done!

thereseanders commented on NAMESPACE:
```diff
@@ -100,3 +111,4 @@ export(user_exists)
 export(watch_repo)
 export(with_pat)
 import(gh)
+importFrom(magrittr,"%>%")
```
done (see above)

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,909 @@
+# Helper function for Latin square
+g = function(j, n) {
```
Changed to `latin_square()`

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,909 @@
+#'
+#' @export
+#'
+peer_create_roster = function(m,
```
Changed to `n_rev`. Also changed in `peer.R` vignette

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,909 @@
+
+#' Assign file to reviewers
+#'
+#' `peer_assign` adds files from authors' repositories to review repositories. It also creates an issue in the reviewers' repositories informing them that the review files are available and creating links to the relevant documents.
```
Changed.

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,909 @@
+#'
+#' `peer_create_form_review` creates blank feedback forms for reviewers based on the user-specified number of questions.
+#'
+#' @param n Numerical. Number of score fields to be included in .Rmd YAML.
```
Changed

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,909 @@
+#' `peer_create_form_review` creates blank feedback forms for reviewers based on the user-specified number of questions.
+#'
+#' @param n Numerical. Number of score fields to be included in .Rmd YAML.
+#' @param title Character. Title of form, defaults to "Reviewer feedback form."
```
Thank you for catching that!

thereseanders replied:
done!

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,909 @@
+    paste0(purrr::map(
+      1:n,
+      ~ paste0(
+        sprintf("%1$i. Place Question %1$i text here. [max. xxx points]\n\n", .x),
```
Do you mean this?
```r
sprintf("#### %1$i. Place Question %1$i text here. [max. xxx points]\n\n", .x)
```

mine-cetinkaya-rundel replied:
yup

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,909 @@
+#'
+#' `peer_create_form_rating` creates a short feedback form for authors to rate the feedback they got from reviewers.
+#'
+#' @param category Character. Categories to be included in the feedback form, defaults to `c("helpfulness", "accuracy", "fairness")`.
```
Added the sentence to the short description before the parameter section:

"The rating categories are based on Reily, K. and P. Ludford Finnerty,  and L. Terveen (2009): Two Peers Are Better Than One: Aggregating Peer Reviews for Computing Assignments is Surprisingly Accurate. In *Proceedings of the ACM 2009 International Conference on Supporting Group Work*. GROUP’09, May 10–13, 2009, Sanibel Island, Florida, USA."

thereseanders commented on vignettes/instructions_students.Rmd:
```diff
@@ -0,0 +1,114 @@
+Your instructor will add you to the ``r params$org`` organization on Github. Once you are added, you will receive an invitation to join to the email address you specified when creating your GitHub account. Please accept the invitation to join the ``r params$org`` organization.
+
+## Assignment
+Your instructor will create a personal repository for your assignment. It might be called something like ``r paste0(params$org, "/hw1-username")``. Please complete the following steps:
```
Done

mine-cetinkaya-rundel commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+    grepl("^\\.", exclude_extension),
+    ~ sub("^\\.", "", .x)
+  ))
+  path[!purrr::map_lgl(path,
```
Change above should remove the need for this, or simplify it greatly.

thereseanders replied:
Changed to 
```r 
  path[!purrr::map_lgl(path,
                       function(path) {
                         any(purrr::map_lgl(rx_exclude_pattern, ~ grepl(.x, path)))
                         })]
```

mine-cetinkaya-rundel commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+  arg_is_chr(path, allow_null = TRUE)
+  arg_is_lgl_scalar(overwrite)
+
+  prefix_rev = format_rev(prefix, suffix)$prefix_rev
```
Instead of `$` use double square brackets to avoid partial matching. (This might be good to check for throughout.)

thereseanders replied:
Changed throughout.

mine-cetinkaya-rundel commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+               if (!is.null(local_path)) {
+                 file_status = fs::file_exists(local_path)
+
+                 if (file_status) {
```
To avoid nesting, which will generally improve readability, if the file doesn't exist, throw an error with the relevant message from below.

If you get past that line, you can assume the file exists, so you don't need to nest subsequent lines of code.

thereseanders replied:
Done!

mine-cetinkaya-rundel commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+                 file_status = fs::file_exists(local_path)
+
+                 if (file_status) {
+                   if (check_rmd) {
```
`if(check_rmd)` and the subsequent if statement can be combined with an and statement.

thereseanders replied:
Changed to
```r 
if (check_rmd & (tolower(fs::path_ext(local_path)) == "rmd"))
```

mine-cetinkaya-rundel commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+
+                 if (file_status) {
+                   if (check_rmd) {
+                     if (!grepl("\\.[rR]md$", fs::path_file(local_path))) {
```
`fs::path_file()` is not necessary here.

Also regex isn't needed here. Use `fs::path_ext()` to extract the file extension, apply `tolower()` to the result, and you can check if that character string is equal to `"rmd"`.

``` r
library(fs)

x <- "hello.Rmd"
"rmd" == tolower(fs::path_ext(x))
#> [1] TRUE
```

thereseanders replied:
Changed to 
```r
if (check_rmd & (tolower(fs::path_ext(local_path)) == "rmd"))
```

mine-cetinkaya-rundel commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+  } else if (!res$result) {
+    usethis::ui_stop("Cannot locate file: {usethis::ui_value(roster)}")
+  } else if (res$result) {
+    rdf = suppressMessages(readr::read_csv(roster))
```
readr should go into imports since we're using `read_csv`

thereseanders replied:
Thank you for catching that!

mine-cetinkaya-rundel commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+                                           check_rmd = TRUE)
+
+  out = purrr::map_df(unique(rdf$author),
+                      function(a) {
```
Use a more useful parameter name, e.g. author.

thereseanders replied:
replaced `a` with `author`

mine-cetinkaya-rundel commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+                          )
+                        }
+
+                        # 2. Mirror files to reviewers
```
"Copy" instead of "Mirror" in the comment

thereseanders replied:
Done

mine-cetinkaya-rundel commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+  arg_is_chr_scalar(message, branch, allow_null = TRUE)
+  arg_is_lgl(overwrite)
+
+  out = purrr::map_dfr(target_repo,
```
In places like this where there are nested maps happening, can you add a comment that says `# TODO: Review to simplify nested mapping` so that we can come back to them later. This doesn't need to hold up the pull request, but is worth revisiting and the tagging will help. Also open an issue in the repo that mentions the comment tag.

thereseanders replied:
Done!

mine-cetinkaya-rundel commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+    added = FALSE
+    commit_sha = NA
+  } else if (!is.null(res) & succeeded(res)) {
+    mode = 100644
```
What is this mode? Can you add a comment next to it saying what it is, maybe with a link to where the info is on the GitHub API.

thereseanders replied:
This is the mode for a file blob, see [documentation of the API](https://developer.github.com/v3/git/trees/#create-a-tree). 

The reason to enter them manually is to reduce calls to the API. The object returned by the API as the output of a `PUT` call does not contain information on `mode` and `type`. The `GET` command from the Trees API returns this information, however that would require an additional call. 

Adding the `mode` and `type` to the output tibble is not necessary. We could just leave `mode = NA` and `type = NA` if we did not want to assume that the file added via `peer_add_content()` is a file blob.

thereseanders replied:
Changed to 
```r
    mode = NA
    type = "blob"
```

rundel replied:
Sounds good, can you put a comment to that effect in the code?

thereseanders replied:
Added the following comment:
```r
    # mode = 100644 & type = "blob" is added manually to reduce API calls
    # mode = 100644 is the mode for a blob, see documentation
    # https://developer.github.com/v3/git/trees/#create-a-tree
    # Needed to accurately track files present on repositories
    # Assumes that only file blobs are added to repositories
```

mine-cetinkaya-rundel commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+                                                              branch = branch)
+
+                        if (length(content_review) > 0) {
+                          repo_files_r_patch = unique(rbind(repo_files_r,
```
It's unclear what's happening here, can you add comments and/or temporary objects?

thereseanders replied:
I added comments to all instances where *_patch objects are created to keep track of files present on reviewer repositories manually to reduce API calls.

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+                              user,
+                              seed = 12345,
+                              write_csv = TRUE) {
+  stopifnot(is.numeric(n_rev))
```
Added following two functions

```r 
arg_is_int = function(..., allow_null = FALSE) {
  handle_arg_list(
    ...,
    tests = function(name, value) {
      if (!(is.integer(value) | (is.numeric(value) && value%%1 == 0)) | (is.null(value) & allow_null))
        usethis::ui_stop("Argument {usethis::ui_value(name)} must be a whole positive number.")
    }
  )
}

arg_is_int_scalar = function(..., allow_null = FALSE, allow_na = FALSE) {
  arg_is_int(...)
  arg_is_scalar(..., allow_null = allow_null, allow_na = allow_na)
}
```

rundel replied:
Current code allows for either + or - integers but the error message mentions only positive integers. We can either add an addition argument to the functions or add another function like `arg_is_pos_int` arg_is_pos_int_scalar`.

thereseanders replied:
Good catch, thank you! Changed to `arg_is_pos_int` and `arg_is_pos_int_scalar`. The code below excludes zero as well, which is what we want for this use case.

```r 
arg_is_pos_int = function(..., allow_null = FALSE) {
  handle_arg_list(
    ...,
    tests = function(name, value) {
      if (!(is.integer(value) | (is.numeric(value) && value > 0 && value%%1 == 0) | (is.null(value) & allow_null)))
        usethis::ui_stop("Argument {usethis::ui_value(name)} must be a whole positive non-zero number.")
    }
  )
}
arg_is_pos_int_scalar = function(..., allow_null = FALSE, allow_na = FALSE) {
  arg_is_pos_int(..., allow_null = allow_null)
  arg_is_scalar(..., allow_null = allow_null, allow_na = allow_na)
}
```

mine-cetinkaya-rundel replied:
👍

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+  stopifnot(is.numeric(n_rev))
+  stopifnot(is.numeric(seed))
+  arg_is_chr(user)
+  stopifnot(length(user) > 1, n_rev > 0, n_rev < length(user))
```
Replaced with the following

```r
  if (!(length(user) > 1)) {
    usethis::ui_stop("{usethis::ui_field('user')} must contain more than one user name.")
  }
  if (!(n_rev > 0)) {
    usethis::ui_stop("{usethis::ui_field('n_rev')} must be at least 1.")
  }
  if (!(n_rev < length(user))) {
    usethis::ui_stop("{usethis::ui_field('n_rev')} must be smaller than the number of users in {usethis::ui_field('user')}.")
  }
```

rundel replied:
This looks good, some of this might be able to be wrapped into some modications to the `arg_is_*` functions but that is a task for another day.

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+#'
+peer_roster_create = function(n_rev,
+                              user,
+                              seed = 12345,
```
Added the following
```r
  if (is.null(seed)) {
    seed = sample(1:1e+06, 1)
    usethis::ui_warn("No seed was supplied. Using randomly sampled seed {usethis::ui_value(seed)}")
  }
```

thereseanders replied:
Added
```r
  if (is.null(seed)) {
    seed = sample.int(1e+06, 1L)
    usethis::ui_warn("No seed was supplied. Using randomly sampled seed {usethis::ui_value(seed)}")
  }
``

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+  arg_is_chr(user)
+  stopifnot(length(user) > 1, n_rev > 0, n_rev < length(user))
+
+  set.seed(seed)
```
Thank you so much for pointing this out. Changed to
```r 
j = withr::with_seed(seed, sample(2:length(user), n_rev))
```

And later
```r
  user_random = paste0("aut", withr::with_seed(seed, sample(1:length(user), length(user))))
```

rundel replied:
I think we want to generate both of these within the same withr::with_seed block, the current code is semi-equivalent to running set.seed() before each line which isn't what we want. (Not actually sure why this is bad, but it seems like it should be avoided)

thereseanders replied:
Changed to

```r
  withr::with_seed(seed, {
    j = sample(2:length(user), n_rev)
    res = purrr::map(j, ~ latin_square(.x, length(user)))
    
    # Randomizing user names to avoid clustering
    user_random = paste0("aut", sample(1:length(user), length(user)))
  })
```

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+  df_sort = data.frame(user = user,
+                       user_random = as.character(user_random))[order(as.numeric(sub("[aA-zZ]+", "", user_random))),]
+
+  res_df = stats::setNames(data.frame(df_sort,
```
Thank you for this suggestion! Changed the code to:

```r
  df_sort = tibble::tibble(user = user,
                           user_random = user_random)[order(as.numeric(sub("[aA-zZ]+", "", user_random))),]

  res_df = stats::setNames(cbind(df_sort,
                                 purrr::map_dfc(res, ~ tibble::tibble(df_sort[['user_random']][.x]))),
                           c("user", "user_random", purrr::map_chr(1:n_rev, ~ paste0("rev", .x))))
```

rundel replied:
`purrr::map_chr(1:n_rev, ~ paste0("rev", .x))` can just be `paste0("rev", 1:n_rev)`, the latter should be fast as well (not that this is performance critical code)

thereseanders replied:
Changed to `paste0("rev", 1:n_rev))`.

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+peer_roster_create = function(n_rev,
+                              user,
+                              seed = 12345,
+                              write_csv = TRUE) {
```
Added parameter `dir = getwd()` to the function and parameter description:

```r
#' @param dir Character. Directory where the peer review roster will be written if `write_csv = TRUE`, defaults to current working directory.
```

Also added the following code:
```r
  if (!dir.exists(dir)) {
    dir = getwd()
    usethis::ui_warn("Directory {usethis::ui_value(dir)} does not exist. Saving to working directory instead.")
  }
```
...
```r
    readr::write_csv(res_df, glue::glue("{dir}/{fname}"))
    usethis::ui_done("Saved file {usethis::ui_value(fname)} to {usethis::ui_value(dir)}.")
```

rundel replied:
Checking for the directory is a good idea, I think I would prefer to throw an error in the case where the directory does not exist. CRAN in particular can be picky about writing files in places where the user didn't explicitly ask for something.

thereseanders replied:
Changed documentation: `#' @param dir Character. Directory where the peer review roster will be written if `write_csv = TRUE`.`

Also changed code to
```r
  if (!dir.exists(dir)) {
    usethis::ui_stop(
      "Directory {usethis::ui_value(dir)} does not exist."
    )
  }
```

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+    readr::write_csv(res_df, fname)
+    usethis::ui_done("Saved file {usethis::ui_value(fname)} to working directory.")
+  } else {
+    tibble::as_tibble(purrr::modify_if(res_df, is.factor, as.character))
```
Done!

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+                                      }),
+                           c("user", "user_random", purrr::map_chr(1:n_rev, ~ paste0("rev", .x))))
+
+  res_df = res_df[order(res_df$user_random),]
```
Added the following
```r
  attr(res_df, "seed") <- seed
  attr(res_df, "user") <- user
```

rundel replied:
Looks good

thereseanders commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+}
+
+# Reads roster file
+peer_roster_read = function(roster) {
```
Done

thereseanders commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+
+# Reads roster file
+peer_roster_read = function(roster) {
+  res = purrr::safely(fs::file_exists)(roster)
```
Changed the function based on your suggestion and added an additional error message if a file could not successfully be read (for example if a file other than a .csv or a broken .csv was supplied). 
``` r
peer_roster_process = function(roster) {

  is_df = is.data.frame(roster)
  is_chr = is.character(roster)

  if (is_df) {
    rdf = tibble::as_tibble(purrr::modify_if(roster, is.factor, as.character))
  } else if (is_chr) {
    if (fs::file_exists(roster)) {
      res = purrr::safely(readr::read_csv)(roster)
      status_msg(res,
                 fail = "Cannot read {usethis::ui_field('roster')}.")
    } else {
      usethis::ui_stop("Cannot locate file: {usethis::ui_value(roster)}")
    }
  } else {
    usethis::ui_stop("{usethis::ui_field('roster')} must be a data.frame or .csv file.")
  }
  rdf
}
```

rundel replied:
Looks good, I dont think the tmp object is needed here - since the if / else will return things. 

```r
   if (fs::file_exists(roster)) {
      res = suppressMessages(purrr::safely(readr::read_csv)(roster))
      status_msg(res,
                 fail = "Cannot read {usethis::ui_field('roster')}.")
      if (succeeded(res))
        tmp = res[["result"]]
    } else {
      usethis::ui_stop("Cannot locate file: {usethis::ui_value(roster)}")
    }
```

can be similarly simplified to

```r
   if (!fs::file_exists(roster)) {
    usethis::ui_stop("Cannot locate file: {usethis::ui_value(roster)}")
  }
  readr::read_csv)(roster))
```

I think, if the roster fails to read we can let readr's error handling take care of it.

thereseanders replied:
Changed to the version below. I kept the `tmp` object because it is used as the input to `rdf = peer_roster_check(tmp)` in the next line. 

```r 
  if (is_df) {
    tmp = tibble::as_tibble(purrr::modify_if(roster, is.factor, as.character))
  } else if (is_chr) {
    if (!fs::file_exists(roster))
      usethis::ui_stop("Cannot locate file: {usethis::ui_value(roster)}")
    tmp = readr::read_csv(roster)
  } else {
    usethis::ui_stop("{usethis::ui_field('roster')} must be a data.frame or .csv file.")
  }
```

thereseanders commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+    rdf = suppressMessages(readr::read_csv(roster))
+  }
+
+  rdf
```
That is a great suggestion!

I added a `ui_stop` statement that throws an error if the roster is not in the expected format.

```r 
  rdf = peer_roster_check(tmp)
  if (!is.null(rdf)) {
    rdf
  } else {
    usethis::ui_stop("Please supply a peer review roster in the expected format (see reference for `peer_roster_create()`).")
  }
```

rundel replied:
What about

```r
  rdf = peer_roster_check(tmp)
  if (is.null(rdf)) {
    usethis::ui_stop(
      "Please supply a peer review roster in the expected format (see reference for `peer_roster_create()`)."
    )
  }
  rdf
```

thereseanders replied:
Thank you for this suggestion! Done!

thereseanders commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+  arg_is_chr_scalar(org, prefix, suffix, prefix_rev, suffix_rev)
+
+  rdf = peer_roster_read(roster)
+  peer_roster_check(rdf)
```
Changed to `rdf = peer_roster_process(roster)`

thereseanders replied:
New version of `peer_roster_check()`


```r 
peer_roster_check = function(roster) {

  # Check user and user_random colnames
  val = c("user", "user_random")
  passed1 = purrr::map_lgl(val,
              function(val) {
                if (!(val %in% names(roster))) {
                  usethis::ui_oops("{usethis::ui_field('roster')} must contain column {usethis::ui_value(val)}")
                  FALSE
                } else {
                  TRUE
                }
              })

  # Check rev* colname(s)
  val2 = "rev*"
  passed2 = TRUE
  if (!(any(grepl("^rev[0-9]+$", names(roster))))) {
    usethis::ui_oops(
      "{usethis::ui_field('roster')} must contain at least one column {usethis::ui_value(val2)}, where '*' denotes a reviewer number."
    )
    passed2 = FALSE
  }

  if (!all(c(passed1, passed2))) {
    NULL
  } else {
    roster
  }
}
```

thereseanders commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+  peer_roster_check(rdf)
+
+  out = purrr::map_dfr(rdf[['user']],
+                       function(y) {
```
Moved to function `peer_roster_format_cols`

thereseanders commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+                         tibble::tibble(
+                           author = y,
+                           author_random = as.character(rdf[rdf$user == y, 'user_random']),
+                           repo_a = as.character(glue::glue("{org}/{prefix}{y}{suffix}")),
```
Added function `glue_chr` as per your suggestion. Not wrapping the `glue::glue()` statements into `as.character()` here causes warning messages:  
`In bind_rows_(x, .id) :
  Vectorizing 'glue' elements may not preserve their attributes`

thereseanders commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+                   description = desc
+                 )
+
+                 if (verbose) {
```
The API returns a 422 Unprocessable Entry error if a label has previously been applied to a repository, see #70. 

The new default is for `peer_init()` not to give information about (un)successful label creation, as to not confuse users. However, users can request information by setting `verbose = TRUE` (summarizes label creation) and/or `show_label_result = TRUE` (outputs tibble). 

```r
peer_issue_label_create = function(repo) {
  arg_is_chr(repo)

  name = c(":pencil: Complete review", ":mag: Inspect review")
  color = c("ff004d", "ffb400")
  desc = c("Please complete review", "Please inspect and rate review")

  purrr::pmap_dfr(list(name, color, desc),
               function(name, color, desc) {
                 res = purrr::safely(github_api_label_create)(
                   repo = repo,
                   name = name,
                   color = color,
                   description = desc
                 )

                 if (succeeded(res)) {
                   created = TRUE
                   error = NA
                 } else {
                   created = FALSE
                   error = res[["error"]][["headers"]][["status"]]
                 }
                tibble::tibble(
                  repo = repo,
                  label = name,
                  created = created,
                  error = error
                )
               })
}


peer_issue_label_apply = function(org,
                                  repo = NULL,
                                  filter = NULL,
                                  exclude = FALSE,
                                  verbose = FALSE,
                                  show_label_result = FALSE) {
  arg_is_chr_scalar(org)
  arg_is_chr_scalar(filter, allow_null = TRUE)
  arg_is_chr(repo, allow_null = TRUE)
  arg_is_lgl(exclude, verbose, show_label_result)

  if (is.null(repo)) {
    repo = org_repos(
      org = org,
      filter = filter,
      exclude = exclude,
      full_repo = TRUE
    )
  }

  usethis::ui_info("Applying labels: This might take a moment...")
  out = purrr::map_dfr(repo, ~ peer_issue_label_create(.x))

  # Note that the 422 error likely means that the label already exists.
  if (verbose)
    peer_issue_label_apply_msg(out)

  if (show_label_result)
    out

}

peer_issue_label_apply_msg = function(label_df) {
  purrr::walk(unique(label_df[["label"]]),
              function(x){

                repo_label_yes = label_df[["repo"]][label_df[["label"]] == x & label_df[["created"]]]
                repo_label_no = label_df[["repo"]][label_df[["label"]] == x & !label_df[["created"]]]

                if (length(repo_label_yes) > 0)
                  usethis::ui_done("Created label {usethis::ui_value(x)} for {usethis::ui_value(length(repo_label_yes))} repositories.")

                if (length(repo_label_no) > 0)
                  usethis::ui_oops("Failed to create label {usethis::ui_value(x)} for {usethis::ui_value(length(repo_label_no))} repositories. The label may already exist. Re-run with `show_label_result = TRUE` for more information.")

              })
}
```

thereseanders commented on R/peer.R:
```diff
@@ -0,0 +1,989 @@
+                       local_path_review = NULL,
+                       prefix = "",
+                       suffix = "",
+                       exclude_extension = c("gitignore", "md", "Rhistory", "Rproj", "html"),
```
Changed to `exclude pattern`. 

**For ".Rproj"**: Is it correct that we want to exclude regex `"^\\.Rproj$"` rather than regex `"^.*\\. Rproj $"`?

rundel replied:
Yeah, that glob should be `*.Rproj`

thereseanders replied:
Thank you! Changed to `exclude_pattern = c(".gitignore", ".Rhistory", "*.Rproj", "*.html", "*.md", "*.pdf")`

thereseanders commented on R/util_peer.R:
```diff
@@ -0,0 +1,676 @@
+
+  path = repo_files[['path']][repo_files[['type']] == "blob" &
+                           !grepl("/", repo_files[['path']])]
+  exclude_extension = unlist(purrr::map_if(
```
Changed to 

```r
  rx_exclude_pattern = utils::glob2rx(exclude_pattern)
  path[!purrr::map_lgl(path,
                       function(path) {
                         any(purrr::map_lgl(rx_exclude_pattern, ~ grepl(.x, path)))
                         })]
```

thereseanders commented:
- Added comments about complex functions staying together to reduce API calls.
- Changed `_r` to` _rev` and `_a` to `_aut` throughout. 
- Added TODO where appropriate.
